joomla/joomla-cms
# Implement invisible Recaptcha

Pull Request for Issue #14565 and following up #14583 and #16599 et al.

### Summary of Changes
This PR implements the possiblity to use the Invisible reCAPTCHA using a separate plugin. Besides implementing the basic functionality to make invisible reCAPTCHA working, there has been made some smaller architectural changes to the whole captcha process (e.g. using Exception instead of JError for errors) and the reCAPTCHA PHP library from Google has been updated, alongside with some smaller changes in the old reCAPTCHA (V1 und V2) plugin.

As this is one of my first PRs for Joomla!, please dont hesitate to make suggestions of how to improve it.

### Testing Instructions
1) Create a site- and secret-key at https://www.google.com/recaptcha/admin for an invisible reCAPTCHA (reCAPTCHA V2 keys will *not* work with Invisible reCAPTCHA)
1) Configure the Invisible reCAPTCHA plugin in the Joomla! backend, entering at least the site- and secret-key
1) Configure the Invisible reCAPTCHA plugin to be used as captcha plugin (System -> Global Configuration -> Default Captcha)
1) Open up a form in the frontend where a captcha is enabled (e.g. by creating a menulink to a contactform)
1) Notice the reCAPTCHA logo on the right side (see screenshot further down) which indicates that the invisible reCAPTCHA has been executed.

We also need to make sure that the reCAPTCHA plugin (V1 and V2) still is working correctly (as there has been made some bigger changes to the code by me), although I have tested this toroughly already, but I would be very glad if other people could take a look at this too, I don't want to take chances that this PR breaks something.

Additionally, as this PR changes the rendering of form fields for the registration form (this was needed as the registration form was using its own rendering of form fields, which caused issues with the invisible reCAPTCHA), we need to make sure the registration form for users is working as expected. So we need to test the registration form in resprect to custom fields of users etc, and whether they get displayed correctly now.

### Expected result
![image](https://user-images.githubusercontent.com/2082625/30965998-9662bb34-a457-11e7-87ec-9c44538a2a8f.png)
